doc/adr/0004-marconi-initiative.rst

```diff
@@ -0,0 +1,100 @@
+
+=================
+Indexing solution
+=================
+
+The indexing solution has the following basic requirements: it needs to deal with rollbacks as elegantly as possible and provide a way to compromise between memory, disk and CPU usage.
+
+On the Cardano blockchain, there are frequent rollbacks, but they can only span a maximum of 2160 blocks (and most of them are < 10 blocks). We call the 2160 number the security parameter K (and we denote it by 'K' henceforth).
+
+Indexers are a store which is updated by events created from each block. The problem introduced by rollbacks is that we need to undo all state changes when a rollback occurs.
+
+We opted for a design where we keep K blocks in memory as the list of events that are fed into the function that merges them into the persisted state and persist them once they go beyond the K limit.
+
+This architectural decision has some desirable effects:
+
+  1. Managing rollbacks is very simple and fast. We drop the events that were rolled back. (no complicated logic is required to undo the projection of the list of events on disk, which we would need if we would want to store everything on disk as fast as possible). 
+  
+  2. Making 'K' configurable makes the design already quite scalable. Developers do not usually need to guard themselves against rollbacks by K blocks so they can choose to only store 10 events in memory thus sacrificing chain integrity for RAM.
+  
+  3. In case of a restart recovery is very simple. We only store fully confirmed transactions so there is nothing to do other than resume operation.
+
+And some less desirable effects:
+
+  1. We must keep K events in memory, which (depending on how large events are) can waste some memory. Our educated guess is that this is a reasonable compromise, but depending on how large events can get that may not be the case for your use case.
+  
+  2. Queries are more complicated as we need to query both events in memory and the state persisted on disk.
+
+=================
+Query and storage
+=================
+
+The indexed data is accessible through queries. There are no constraints on the format of queries or results. Both are identified by a type variable that the indexer exposes and the code for both of them has to be provided by the user.
+
+This is currently the only area where the users can customise the indexer by providing functions that handle queries and storage. Note that the queries have to run on both the in-memory data and the data that is stored.
+
+This allows us to associate any kind of storage type to the indexers, though, right now we are only using SQLite.
+
+==================
+Versioned contents
+==================
+
+We need some sort of versioning or sequencing of events, notifications and queries.
```

Comment:
I am not sure what you mean by this.

Reply:
I rewrote it. Hopefully it's a little more clear now.